File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:

   void helper(vector<int>& arr,vector<vector<int>>& ans,vector<int>& temp, int target,int idx ){
    if(idx == arr.size() ||  target < 0) return;
    if(target == 0){
            ans.push_back(temp);
            return;
        }
  

    temp.push_back(arr[idx]);
    helper(arr,ans,temp, target-arr[idx], idx);
    temp.pop_back();
    helper(arr,ans,temp, target, idx+1);
   }
    vector<vector<int>> combinationSum(vector<int>& arr, int target) {
        vector<vector<int>>ans; vector<int>temp;
        helper(arr,ans,temp,target,0);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {

    void help(int idx ,int[] candidates, int target, List<List<Integer>> ans, List<Integer> temp){
         if(target<0 || idx == candidates.length){
            return;
        }
        if(target == 0){
            ans.add(new ArrayList<>(temp));
            return;
        }
       temp.add(candidates[idx]);
        help(idx,candidates, target-candidates[idx], ans , temp);
        temp.remove(temp.size()-1);
        help(idx+1,candidates, target, ans , temp);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
       List<List<Integer>> ans = new ArrayList<>();
       List<Integer> temp = new ArrayList<>();
       help(0,candidates,target,ans,temp);
       return ans;
    }
}